tests/evm/test_extcodecopy.py

```diff
@@ -0,0 +1,167 @@
+import pytest
+from itertools import chain
+from zkevm_specs.copy_circuit import verify_copy_table
+from zkevm_specs.evm import (
+    verify_steps,
+    Tables,
+    Block,
+    Bytecode,
+    ExecutionState,
+    StepState,
+    Opcode,
+    U64,
+    U160,
+    RLC,
+    CopyCircuit,
+    CopyDataTypeTag,
+    AccountFieldTag,
+    CallContextFieldTag,
+)
+from zkevm_specs.evm.typing import RWDictionary
+from zkevm_specs.util import (
+    rand_fq,
+    keccak256,
+    EXTRA_GAS_COST_ACCOUNT_COLD_ACCESS,
+    GAS_COST_COPY,
+    memory_word_size,
+    memory_expansion,
+)
+
+TESTING_DATA = (
+    (bytes(), True, True, 0x30000, 0x00, 0x00, 54),  # warm account with empty code
+    (bytes(), False, True, 0x30000, 0x00, 0x00, 54),  # cold account with empty code
+    (bytes([10, 40]), True, True, 0x30000, 0x00, 0x00, 54),  # warm account
+    (bytes([10, 10]), False, True, 0x30000, 0x00, 0x00, 54),  # cold account
+)
+
+
+@pytest.mark.parametrize(
+    "code, is_warm, is_persistent, address, src_addr, dst_addr, length", TESTING_DATA
+)
+def test_extcodecopy(
+    code: bytes,
+    is_warm: bool,
+    is_persistent: bool,
+    address: U160,
+    src_addr: U64,
+    dst_addr: U64,
+    length: U64,
+):
+    randomness = rand_fq()
+
+    code_hash = int.from_bytes(keccak256(code), "big")
+
+    next_memory_word_size = memory_word_size(dst_addr + length)
+    _, memory_expansion_cost = memory_expansion(0, dst_addr + length if length else 0)
+    memory_gas_cost = memory_expansion_cost + memory_word_size(length) * GAS_COST_COPY
+
+    gas_cost_extcodecopy = (
+        Opcode.EXTCODECOPY.constant_gas_cost()
+        + memory_gas_cost
+        + (not is_warm) * EXTRA_GAS_COST_ACCOUNT_COLD_ACCESS
+    )
+
+    tx_id = 1
+    call_id = 1
+
+    rw_counter_end_of_reversion = 0 if is_persistent else 9
+    reversible_write_counter = 0
+
+    rw_dictionary = (
+        RWDictionary(1)
+        .stack_read(call_id, 1020, RLC(address, randomness))
+        .stack_read(call_id, 1021, RLC(dst_addr, randomness))
+        .stack_read(call_id, 1022, RLC(src_addr, randomness))
+        .stack_read(call_id, 1023, RLC(length, randomness))
+        .call_context_read(tx_id, CallContextFieldTag.TxId, tx_id)
```

Comment:
```suggestion
        .call_context_read(call_id, CallContextFieldTag.TxId, tx_id)
```

The tests were passing because `tx_id == call_id == 1`, but the call context read/write is populated with the `call_id`. So this should be changed.

Reply:
OOOOH!!! I completelly missed it! 
This brings in the fact that we should probably pseudo-randomize or do something similar for these cases.
As this should never happen to pass by coincidence..

cc: @ed255

Reply:
As @CPerezz commented, in order to prevent this coincidence, I assigned the random values to `tx_id` and `call_id` in [395fc2c](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/257/commits/395fc2c7995cc4bf74873bc107df17e0a09207ec). Is it okay?

```
tx_id = rand_fq()
call_id = rand_fq()
```

Reply:
If there's an easy way to use Seeded randomness would be better. As then the test case errors are reproducible.

Reply:
> If there's an easy way to use Seeded randomness would be better. As then the test case errors are reproducible.

I agree 100%  I would either use a seedable random number generation (with a constant seed), or even just some constant values like `tx_id = 2, call_id = 3`.

I want to note that as far as I remember we have several tests in the specs that use randomly generated values, which I think should be updated to use constant random values.

Reply:
And in general, maybe it's a good idea to go through all the tests and make sure that we always use a `call_id` that is different than the `tx_id` to find these kind of bugs.